### DISCUSSION TITLE
Disabling automatic mapping of columns to PHP types (mysqli driver)

### DISCUSSION THREAD

**Discussion by pwaring** [ADOdb/ADOdb#988] [Q&A] [answered]
I've recently upgraded the ADODB dependency in a legacy application (using ADODB since PHP 4) and this has broken any queries that use bound parameters on the mysqli driver. The reason for this is that older versions of ADODB returned all database columns as strings, and the application assumes that this will be the case and processes them as such (e.g. `if $x['id'] === '1'`). However, at some point ADODB switched to mapping columns to PHP types, e.g. an INT column will be converted to a PHP int. This only happens on queries with bound parameters for some reason, i.e. where the `query` function is called with a second parameter (defaults to `false` if not provided).

Is there a way to disable this automatic mapping? Ideally I would fix the application, but there are hundreds of instances where it assumes strings and no way to fix them gradually - they would have to be done in one go.

**Comment by dregad** [comment, marked as answer]
This is a consequence of ADOdb switching to true prepared statements with bound variables in v5.22.0 (see #655). 

The typing of returned data is standard PHP mysqli driver behavior, we have no control over it. Consider this sample code using native PHP mysqli extension:

```php
$my = new mysqli($host, $user, $password, $database);

// Query execution without parameters
var_dump($my->query('select id, lastname from people where id = 1')->fetch_assoc());
/*
array(2) {
  'id' =>
  string(1) "1"
  'lastname' =>
  string(6) "Turing"
}
*/

// Executing the same query using a prepared statement
$st = $my->prepare('select id, lastname from people where id = ?');
$st->execute([1]);
var_dump($st->get_result()->fetch_assoc());
/*
array(2) {
  'id' =>
  int(1)
  'lastname' =>
  string(6) "Turing"
}
*/
```

Notice the type of the integer `id` column is returned as **string** in the first case, and **int** with the prepared statement.

The mysqli driver does not provide any way to disable this behavior, and so neither does ADOdb.

As a workaround, you could use v5.21.4, which still uses the legacy, ADOdb-emulated bound variables. This will give you time to gradually adapt your application code.

Alternatively, you can try to create a custom driver extending ADOdb mysqli driver, and post-process query execution / fetching to convert everything to string, but that's no small undertaking.

**Comment by pwaring** [comment]
Thanks for the quick reply. I'm not sure this application can be gradually migrated as there doesn't seem to be a way to support both the pre-5.22.0 and post-5.22.0 behaviours simultaneously, and we can't stay on an old version for too long as we will be upgrading PHP soon. I'll have to go away and give this more thought.

**Reply by dregad** [reply]
What I  meant is that you can upgrade to 5.21 now, which should be OK with PHP up to 8.0, then work on adapting your app and move to 5.22 in a second phase.
